### PR TITLE
[FIX] Prevent FS write from draining Resources content

### PR DESCRIPTION
**Pull request by RandomByte**
Previously, when writing to an FS-Adapter, a Resources content stream
would be drained, making it impossible to access a Resources content
again.

With this change, by default a Resources content will be buffered while
draining its content stream.

An optional 'drain' parameter can be supplied to skip buffering to lower
the memory demand of the operation in cases where further access to the
resource content is not required.

Alternatively, an optional 'readOnly' parameter can be supplied to skip
buffering and instead create a new readable stream of the written file.
The file will be written with '444' (read only) permissions. This can
also help reducing memory usage.

Add a pessimistic check for drained content streams.

Refactor FS- and memory adapter unit tests.

Resolves blocker for https://github.com/SAP/ui5-builder/pull/42

### CODE REVIEW

**Comment by RandomByte** [comment]
Ready for review

File: lib/adapters/FileSystem.js [hunk 1]
```diff
@@ -182,15 +199,48 @@ class FileSystem extends AbstractAdapter {
 			fs
```
Comment (matz3) [inline lib/adapters/FileSystem.js]: `await makeDir(...)` ?

File: lib/AbstractReaderWriter.js [hunk 2]
```diff
@@ -25,22 +25,34 @@ class AbstractReaderWriter extends AbstractReader {
 	 * Writes the content of a resource to a path.
 	 *
 	 * @public
-	 * @param {module:@ui5/fs.Resource} resource The Resource to write
+	 * @param {module:@ui5/fs.Resource} resource Resource to write
+	 * @param {Object} [options]
+	 * @param {boolean} [options.readOnly=false] Whether the resource content shall be written read-only
```
Comment (matz3) [inline lib/AbstractReaderWriter.js]: I'm not completely sure whether I have understood the use case and behavior of `readOnly`. Where will this be used?

File: lib/Resource.js [hunk 3]
```diff
@@ -74,16 +85,21 @@ class Resource {
 	 * Gets a buffer with the resource content.
 	 *
 	 * @public
-	 * @returns {Promise<Buffer>} A Promise resolving with a buffer of the resource content.
+	 * @returns {Promise<Buffer>} Promise resolving with a buffer of the resource content.
 	 */
 	getBuffer() {
-		return new Promise((resolve, reject) => {
+		return Promise.resolve().then(() => {
```
[the remark this answers is not in the document]
Comment (RandomByte) [reply lib/Resource.js]: Done

**Comment by RandomByte** [comment]
For future reference: Some more clarification on the issue this PR resolves can be found here: https://github.com/SAP/ui5-builder/pull/42#issuecomment-401191973

**Pull request merged**